in_toto/process.py

```diff
@@ -118,3 +123,97 @@ def run(cmd, check=True, timeout=SUBPROCESS_TIMEOUT, **kwargs):
+
+
+def run_duplicate_streams(cmd, timeout=SUBPROCESS_TIMEOUT):
+  """
+  <Purpose>
+    Provide a function that executes a command in a subprocess and returns its
+    exit code and the contents of what it printed to its standard streams upon
+    termination.
+
+    NOTE: The function might behave unexpectedly with interactive commands.
+
+
+  <Arguments>
+    cmd:
+            The command and its arguments. (list of str, or str)
+            Splits a string specifying a command and its argument into a list
+            of substrings, if necessary.
+
+    timeout: (default see settings.SUBPROCESS_TIMEOUT)
+            If the timeout expires, the child process will be killed and waited
+            for and then subprocess.TimeoutExpired  will be raised.
+
+  <Exceptions>
+    securesystemslib.exceptions.FormatError:
+            If the `cmd` is a list and does not match
+            in_toto.formats.LIST_OF_ANY_STRING_SCHEMA.
+
+    OSError:
+            If the given command is not present or non-executable.
+
+    subprocess.TimeoutExpired:
+            If the process does not terminate after timeout seconds. Default
+            is `settings.SUBPROCESS_TIMEOUT`
+
+  <Side Effects>
+    The side effects of executing the given command in this environment.
+
+  <Returns>
+    A tuple of command's exit code, standard output and standard error
+    contents.
+
+  """
+  if isinstance(cmd, six.string_types):
+    cmd = shlex.split(cmd)
+  else:
+    formats.LIST_OF_ANY_STRING_SCHEMA.check_match(cmd)
+
+  # Use temporary files as targets for child process standard stream redirects
+  # They seem to work better (i.e. do not hang) than pipes, when using
+  # interactive commands like `vi`.
+  stdout_fd, stdout_name = tempfile.mkstemp()
+  stderr_fd, stderr_name = tempfile.mkstemp()
+  try:
+    with io.open(stdout_name, "r") as stdout_reader, \
+        io.open(stderr_name, "r") as stderr_reader:
+
+      # Start child , writing standard streams to temporary files
+      proc = subprocess.Popen(cmd, stdout=stdout_fd,
+          stderr=stderr_fd, universal_newlines=True)
+      proc_start_time = time.time()
+
+      stdout_str = stderr_str = ""
+      stdout_part = stderr_part = ""
+
+      # Read as long as the process runs or there is data on one of the streams
+      while proc.poll() is None or stdout_part or stderr_part:
+
+        # Raise timeout error in they same manner as `subprocess` would do it
+        if (timeout is not None and
+            time.time() > proc_start_time + timeout):
+          proc.kill()
+          proc.wait()
+          raise subprocess.TimeoutExpired(cmd, timeout)
+
+        # Read from child process's redirected streams, write to parent
+        # process's standard streams and construct retuirn values
+        stdout_part = stdout_reader.read()
+        stderr_part = stderr_reader.read()
+        sys.stdout.write(stdout_part)
+        sys.stderr.write(stderr_part)
+        sys.stdout.flush()
+        sys.stderr.flush()
```

Comment:
silly comment, but are all of these IPC calls available on windows too?

Reply:
According to the passing windows tests on appveyor they do.